### DISCUSSION TITLE
ID / Name label option for Filterbox

### DISCUSSION THREAD

**Discussion by durchgedreht** [apache/superset#19184] [Ideas]
**EDIT 2024:** The "Filterbox" filters discussed here have been removed from Superset starting in 4.0.  They have been replaced by Dashboard Native Filters.  Please see [the version of this request which focuses on Native Filters](https://github.com/apache/superset/discussions/26084).
<hr/>
**Is your feature request related to a problem? Please describe.**
In many DBs there's an ID / Name pattern. So the *id* is integer and *name* is String for human readable representation.
When using Filterbox you have to decide to use id field or String field, but ideally you'd like to use the id field for the technical filter and use the name for labelling only.

example table:
<table>
<tr>
<th>id</th>
<th>name</th>
<th>revenue</th>
</tr>
<tr>
<td>1</td>
<td>sports</td>
<td>1000</td>
</tr>
<tr>
<td>2</td>
<td>cars</td>
<td>3000</td>
</tr>
<tr>
<td>3</td>
<td>entertainment</td>
<td>20000</td>
</tr>
<table>

Benefits would be:

- User still sees a list of Strings to include (as if he selected *name*)
- *preselecxt_fiter* option could inject INTEGER values from the outside. This is especially useful from an integration perspective as normally APIs and applications that link directly into superset and want to preselect a filter will work on the ID rather than on the name 
- imagine the name is changed -this would break the 'contract' between the url parameters and the DB values. The argument, it's based on the data is very weak as names are considered dimension data and thus any data tool should rely on the ID field.  

**Describe the solution you'd like**
It would be great to have an option 'verbose column' (or similar) in the filterbox's parameters. That column selection is optional, but would take care about the  ID -> string mapping.  

<img width="390" alt="Screenshot 2021-03-01 at 23 51 10" src="https://user-images.githubusercontent.com/4490840/109569700-1ef0d380-7ae9-11eb-9d3c-8f5135225a87.png">

In the UI filter box a user would just see values like 'sports', 'cars', 'entertainment', but it could be controlled by preselect_filter using IDs 1, 2, 3.
<img width="326" alt="Screenshot 2021-03-02 at 00 02 26" src="https://user-images.githubusercontent.com/4490840/109570645-a9860280-7aea-11eb-82ae-d0111c117858.png">

**Describe alternatives you've considered**
If the 'verbose' field could be entered using a dedicated SQL expression potentially controlled/secured by some JINJA templating, that would maby more technical, but also more flexible.

**Comment by junlincc** [comment]
@durchgedreht thanks for suggesting. we included this item in 2021Q1 roadmap(tentative), but will mostly likely to implement the enhancement on the dashboard native filter, which is currently behind a feature flag.

**Comment by regisb** [comment]
Hi! I enabled dashboard native filters on my instance but did not see this feature. I also could not find a report about this feature in the related PRs:
https://github.com/apache/superset/issues/12148
https://github.com/apache/superset/issues/13591

Can someone please confirm that this feature is still on the roadmap? If yes, what's the corresponding ticket?

**Comment by jhult** [comment]
@junlincc and @suddjian,

Any news on this?

**Comment by danielsmeyer** [comment]
Is there a branch that I could checkout to get started on helping push this forward? @jhult @junlincc @suddjian 
I have some time to get this integrated into the native filter.

**Comment by Eragra3** [comment]
Hello, this seems like a relatively simple feature, but would make a lot of stuff easier.

Is this item still on the roadmap?

**Comment by junlincc** [comment]
Will pick it up, thanks for letting us know!

**Comment by amitmiran137** [comment]
This got enabled in native filters and will not be addressed under filter box which is depracated already

**Comment by mathieudruart** [comment]
@amitmiran137 Hello, do you know which version of superset enable this in native filters ? Thank you !

**Comment by amitmiran137** [comment]
Probably 1.3.2

**Comment by mathieudruart** [comment]
@amitmiran137 I'm using superset 1.3.2 but I can't figure how to do it, maybe I'm missing something ?

**Comment by nytai** [comment]
Reopening this issue as I does not appear to be solved. If I am mistaken could you please link a screenshot or PR so users are aware of how the feature works?

cc @amitmiran137 @junlincc

**Comment by rusackas** [comment]
From what I can tell, this was not built as part of the original Native Filters scope. We might be able to shoehorn it into some upcoming work around Native Filters, as it seems to me like a perfectly reasonable and useful feature addition. That said, it's more of a feature request than a bug report, per se, so I'm tempted to move this to the "Ideas" area of Discussions instead of its current state as an issue. Any objections to the move?

**Comment by villebro** [comment]
I can confirm this is not yet available. In a sense adding this feature isn't very difficult: The `Select` component already supports having different labels and values, and adding it to the filter component is relatively trivial. However, I'm more concerned with the following aspects of this feature:
1) The native filter control panel has already grown way out of proportion, and we'd need to generalize it some more so that it would be more similar to the Explore control panel. I didn't yet check, but I fear adding this dropdown to the Value filter would require yet another if/then/else to the filter config modal.
2) If the user chooses a combination of value and label columns that aren't 1-1 mapped, the user may get very strange behavior. To avoid confusing error states, we would probably need to verify that all value/label pairs are strictly 1-1.

I'm most concerned with 1), but as I haven't worked on this area for a while I'm not sure if this is a valid concern. @michael-s-molina , do you know if we could easily add an optional column label selector to the Value filter without adding more tech debt to the filter config modal?

**Comment by michael-s-molina** [comment]
@villebro We are currently working on stage 1 of the native filters design changes. The primary objective is to remove the old filter box at the end of this stage. I would advise including this requirement in stage 2 analysis where we'll probably refactor the whole module and your concerns will be addressed.

**Comment by villebro** [comment]
Thanks @michael-s-molina , that sounds like a good plan!

**Comment by rvazquezglez** [comment]
I used this command on chrome to remove the elements:

```
Array.from(document.querySelectorAll('td')).filter(el=> el.textContent === '0.00%').map(i => i.parentNode).forEach(elem => elem.parentNode.removeChild(elem))
```

**Comment by almirb** [comment]
I would also like to filter dashboards using IDs instead of Names.  That would be a great feature!

**Comment by zufolo441** [comment]
Hi all, is there some news regarding this feature? thanks.

**Comment by ricardocarrola** [comment]
Any news on this feature ? Thank you!

**Comment by IgorSterkhov** [comment]
more than 2 years and nothing done yet/...

**Reply by michael-s-molina** [reply]
@IgorSterkhov FilterBox [was deprecated](https://github.com/apache/superset/pull/23784) and will be removed in a future major version. [Native filters](https://docs.preset.io/docs/dashboard-filtering) are the replacement for filter boxes and you can use this [tool](https://superset.apache.org/docs/miscellaneous/native-filter-migration/) to migrate your filters.

**Reply by mathieudruart** [reply]
@michael-s-molina as far as I know this feature is still not supported by native filters? Unless I missed something?

**Reply by michael-s-molina** [reply]
Hi @mathieudruart. You're correct. This feature is not supported by native filters yet. I think it would be a good idea to open a new discussion about this feature in the context of the native filters given that there are many differences between the solutions. It seems a  reasonable and valuable feature.

**Reply by polmonso** [reply]
I opened [this](https://github.com/apache/superset/discussions/26084)

**Comment by iamsaso** [comment]
Any updates on this? I have `location_id` and `location_name` and all the tables have `location_id`. ATM I have to join the name to each table to make filtering human friendly.

**Comment by Regonos** [comment]
Hello, what is a status of that topic?

**Comment by rusackas** [comment]
Filterbox has been removed on `master` and will be gone in 4.0. The other feature still isn't built. PRs welcome, or you can pick up the conversation on [this rebooted thread](https://github.com/apache/superset/discussions/26084).

**Comment by ttauzo** [comment]
Any news on this?
I must say that I as astonished that such really basic business feature is not available from the beginning - business users do not use ids usually but labels. 
I tried to think about workaround but I decided i do not plan to add labels to datasets as I have indexes on ids so querying by label would mean changes in indexing in whole db. - so lacking this feature is a show stopper for me... (not sure i get go without it with my project)
I guess I BI tools have dimensions in a form id, label that should be availabe for business users.


Generally from my perspective superset would gain much functionality if filtering was more complex/parametrizable for business users (similar in a form to eg. Jquery builder).

**Comment by sbkim-muhayu** [comment]
Any news on this feature ?

**Comment by OrwaRebootMob** [comment]
I am interested in this feature too. And might be able to convince my company to invest in me adding it.

**Comment by MammadTavakoli** [comment]
Is there a solution for this problem?

**Comment by amitmiran137** [comment]
This is amazing an amazing milestone

**Comment by calebJ-MS** [comment]
Still nothing? This would be such a nice feature

**Comment by leschiffres** [comment]
This would be a great addition, to the tool. Not sure how much complexity it introduces though.
